sdk/core/azure-core/src/http/winhttp/win_http_request.hpp

```diff
@@ -0,0 +1,213 @@
+
+namespace Azure { namespace Core { namespace Http { namespace _detail {
+
+  class WinHttpRequest;
+  /**
+   * @brief An outstanding WinHTTP action. This object is used process asynchronous WinHTTP actions.
+   *
+   * The WinHttpRequest object has a WinHttpAction associated it to convert asynchronous WinHTTP
+   * operations to synchronous operations.
+   *
+   */
+  class WinHttpAction final {
+    friend class WinHttpRequest;
+
+    // Containing HTTP request, used during the status operation callback.
+    WinHttpRequest* m_httpRequest{};
+    std::mutex m_actionCompleteMutex;
+    //    std::condition_variable m_actionCompleted;
+    wil::unique_event m_actionCompleteEvent;
+    DWORD m_stowedError{};
+    DWORD_PTR m_stowedErrorInformation{};
+    DWORD m_bytesAvailable{};
+
+    /*
+     * Callback from WinHTTP called after the TLS certificates are received when the caller sets
+     * expected TLS root certificates.
+     */
+    static void CALLBACK StatusCallback(
+        HINTERNET hInternet,
+        DWORD_PTR dwContext,
+        DWORD dwInternetStatus,
+        LPVOID lpvStatusInformation,
+        DWORD dwStatusInformationLength) noexcept;
+
+    /*
+     * Callback from WinHTTP called after the TLS certificates are received when the caller sets
+     * expected TLS root certificates.
+     */
+    void OnHttpStatusOperation(
+        HINTERNET hInternet,
+        DWORD internetStatus,
+        LPVOID statusInformation,
+        DWORD statusInformationLength);
+
+  public:
+    /**
+     * @brief Create a new WinHttpAction object associated with a specific WinHttpRequest.
+     *
+     * @param request Http Request associated with the action.
+     */
+    WinHttpAction(WinHttpRequest* request)
+        : m_httpRequest(request), m_actionCompleteEvent(CreateEvent(nullptr, TRUE, FALSE, nullptr))
+    {
+      if (!m_actionCompleteEvent)
+      {
+        throw std::runtime_error("Error creating Action Complete Event.");
+      }
+    }
+    /**
+     * @brief WaitForAction - Waits for an action to complete.
+     *
+     * @remarks The WaitForAction method waits until an action initiated by the `callback` function
+     * has completed. It polls with a timeout of milliseconds, checking to see if the provided
+     * context parameter has timed out.
+     *
+     * @param callback - Callback used to initiate an action. Always called in the waiting thread.
+     * @param pingDuration - The time to wait for a ping to complete.
+     * @param context - Context for the operation.
+     *
+     * @returns true if the action completed normally, false if there was an error.
+
+     * @remarks If there is an error, the caller can determine the error code by calling
+     * GetStowedError() and GetStowedErrorInformation()
+     */
+    bool WaitForAction(
+        std::function<void()> callback,
+        Azure::Core::Context const& context,
+        Azure::DateTime::duration const& pingDuration = std::chrono::milliseconds(800));
```

Comment:
Why is it defaulted to 800 ms? Is that supposed to be the same value as `RetryDelay`:
https://github.com/Azure/azure-sdk-for-cpp/blob/be4cb8f51639be2a36a3f47e014ff4508cc85f69/sdk/core/azure-core/inc/azure/core/http/policies/policy.hpp#L93

Reply:
I needed a polling duration as some number between 0 and 10 minutes (10 minutes is obviously too long, 0 is too short for HTTP), 800ms seemed like a reasonable value because it tracks the retry delay.

Reply:
This would be worth stating as a comment :)